src/tracing.c

```diff
@@ -0,0 +1,204 @@
+/****************************************************************************
+**
+**  This file is part of GAP, a system for computational discrete algebra.
+**
+**  Copyright of GAP belongs to its developers, whose names are too numerous
+**  to list here. Please refer to the COPYRIGHT file for details.
+**
+**  SPDX-License-Identifier: GPL-2.0-or-later
+**
+**  This file contains functionality for tracing operations
+**
+*/
+
+#include "tracing.h"
+
+#include "bool.h"
+#include "integer.h"
+#include "lists.h"
+#include "modules.h"
+#include "plist.h"
+#include "precord.h"
+#include "records.h"
+
+// This file contains functionality which stores how often internal operations
+// (like +, ^, -) are executed on each TNUM.
+
+
+static Obj RecordedStats;
+
+// Report the 1 argument operation 'name' was applied to 'op'
+void ReportWrappedOperation1(const char * cname, Obj op)
+{
+    UInt name = RNamName(cname);
+    if (!ISB_REC(RecordedStats, name)) {
+        Obj list = NEW_PLIST(T_PLIST, 0);
+        ASS_REC(RecordedStats, name, list);
+    }
+    Obj list = ELM_REC(RecordedStats, name);
+
+    UInt tnam = TNUM_OBJ(op) + 1;
+    Obj  val = ELM0_LIST(list, tnam);
+    if (!val) {
+        val = INTOBJ_INT(0);
+    }
+
+    UInt8 intval = Int8_ObjInt(val);
+    intval++;
+    val = ObjInt_Int8(intval);
```

Comment:
or just `val = SUM(val, INTOBJ_INT(1))` ? Of course there is a performance penalty; but the above code is far from optimal either, and this replacement at least is easy to understand. If performance truly is a concern (there are some good arguments to be made to this end, of course), then I'd instead restrict to INTOBJ, and then one can perform this much faster. But then `ReportWrappedOperation1` is an extern function, and can't be inlined, so this whole setup is far from being efficient anyway. I mean, just accessing a record for each of these operations is a severe performance hit. So, if we say "this simply makes things quite a bit slower, accept it, I'd prefer the simpler code. If we say "we should try to make this as fast as possible", I'd ditch the whole code, and go completely different about it: create C int arrays storing these counts for each unary/binary operation (as 
part of the `DEFINE_OP_WRAPPER1/2` macros, and store everything in those.

Reply:
There's a much more boring reason: SUM uses SumFuncs, which is traced, which would cause an infinite loop. I'll add a comment to say why I'm avoiding SUM here.